evap/results/tools.py

```diff
@@ -171,42 +211,91 @@ def _calculate_results_impl(course):
+
+    contributor_averages = []
+    contributor_distributions = []
+    course_grade_questions_average = None
+    course_grade_questions_distribution = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
+    course_non_grade_questions_average = None
+    course_non_grade_questions_distribution = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
+
+    for contributor in grouped_results:
+        if contributor:
+            # collect the grade question average, non-grade question average and the distributions for each contributor
+            grade_questions_average, grade_questions_distribution, non_grade_questions_average, non_grade_questions_distribution = \
+                calculate_average_grade_and_distribution_for_contribution_results(grouped_results[contributor])
+            contributor_averages.append(mix(grade_questions_average, non_grade_questions_average, settings.CONTRIBUTOR_GRADE_QUESTIONS_PERCENTAGE))
+            contributor_distributions.append(mix_distributions(grade_questions_distribution, non_grade_questions_distribution, settings.CONTRIBUTOR_GRADE_QUESTIONS_PERCENTAGE))
+        else:
+            # calculate the grade question average, non-grade question average and the distributions for the course
+            course_grade_questions_average, course_grade_questions_distribution, course_non_grade_questions_average, course_non_grade_questions_distribution = \
+                calculate_average_grade_and_distribution_for_contribution_results(grouped_results[contributor])
+
+    # calculate the average grade and distribution over all contributors
+    contributors_average = avg(contributor_averages)
+    contributors_distribution = avg_distribution(contributor_distributions)
+
+    # calculate to which percentage each group influences the total grade
+    contributions_percentage, course_grade_questions_percentage, course_non_grade_questions_percentage = \
+        get_grade_percentages(contributors_average is not None, course_grade_questions_average is not None, course_non_grade_questions_average is not None)
+
+    # calculate average grade
+    average_grade = contributions_percentage * (contributors_average or 0) + \
+                    course_grade_questions_percentage * (course_grade_questions_average or 0) + \
+                    course_non_grade_questions_percentage * (course_non_grade_questions_average or 0)
+
+    # calculate average distribution
+    distributions = [
+        (contributors_distribution, contributions_percentage),
+        (course_grade_questions_distribution, course_grade_questions_percentage),
+        (course_non_grade_questions_distribution, course_non_grade_questions_percentage),
+    ]
+    average_distribution = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
+    for distribution, percentage in distributions:
+        distribution_sum = sum(distribution.values())
+        if distribution_sum > 0:
+            for answer, count in distribution.items():
+                average_distribution[answer] += percentage * (count / distribution_sum)
```

Comment:
right now it looks to me like the contributors_distribution contains percentages, and the course distributions contain counts. i think that should be changed, but that's kind of part of the other comments.